plutus-tx-plugin/executables/profile/Main.hs

```diff
@@ -24,12 +42,120 @@ import qualified PlutusCore.Default        as PLC
 
 import           Control.Lens.Combinators  (_2)
 import           Control.Lens.Getter       (view)
+import           Data.List                 (intercalate)
+import           Data.Maybe                (fromJust)
 import           Data.Proxy                (Proxy (Proxy))
-import           Data.Text                 (Text)
+import qualified Data.Text                 as T
+import           Data.Time.Clock           (NominalDiffTime, UTCTime, diffUTCTime)
 import           Prettyprinter.Internal    (pretty)
 import           Prettyprinter.Render.Text (hPutDoc)
 import           System.IO                 (IOMode (WriteMode), withFile)
 
+data StackFrame
+  = MkStackFrame
+  { -- | The variable name.
+    varName           :: T.Text,
+    -- | The time when it starts to be evaluated.
+    startTime         :: UTCTime,
+    -- | The time spent on evaluating the functions it called.
+    timeSpentCalledFn :: NominalDiffTime
+  }
+  deriving (Show)
+
+-- | Write the time log of a program to a file in
+-- the plutus-tx-plugin/executables/profile/ directory.
+writeLogToFile ::
+  ToUPlc a PLC.DefaultUni PLC.DefaultFun =>
+  -- | Name of the file you want to save it as.
+  FilePath ->
+  -- | The program to be profiled.
+  [a] ->
+  IO ()
+writeLogToFile fileName values = do
+  let filePath = "plutus-tx-plugin/executables/profile/"<>fileName<>".timelog"
+  log <- T.intercalate "\n" . view _2 <$> (rethrow $ runUPlcProfileExec values)
+  writeFile filePath (T.unpack log)
+  processed <- processLog filePath
+  writeFile (filePath<>".stacks") (printProcessedLog processed)
+  pure ()
+
+data ProfileEvent =
+  MkProfileEvent UTCTime Transition T.Text
+
+data Transition =
+  Enter
+  | Exit
+
+data StackTime =
+  MkStackTime [T.Text] Double
```

Comment:
Worth a comment about the function names, in particular that they're backwards. And you could say that this basically corresponds to one of the "folded" flamegraph lines.